src/test/java/com/treasuredata/client/TestTDHttpClient.java

```diff
@@ -78,4 +93,127 @@ public void deleteMethodTest()
+                        return new HttpContentResponse(response, new byte[] {}, "", "");
+                    }
+                    case 2: {
+                        secondRequestNanos.set(System.nanoTime());
+                        HttpResponse response = new HttpResponse(request, listeners)
+                                .status(200);
+                        return new HttpContentResponse(response, body, "plain/text", "UTF-8");
+                    }
+                    default:
+                        throw new AssertionError();
+                }
+            }
+        });
+
+        assertThat(requests.get(), is(2));
+        assertThat(resp.getStatus(), is(200));
+        assertThat(resp.getContent(), is(body));
+
+        long delayNanos = secondRequestNanos.get() - firstRequestNanos.get();
+        assertThat(delayNanos, Matchers.greaterThanOrEqualTo(SECONDS.toNanos(retryAfterSeconds)));
+    }
+
+    @Test
+    public void failOn429_TimeLimitExceeded()
+            throws Exception
+    {
+        client = TDClient.newBuilder()
+                .setRetryMaxIntervalMillis(1000)
+                .setRetryLimit(3)
+                .build()
+                .httpClient;
+
+        final long retryAfterSeconds = 4711;
+        final AtomicInteger requests = new AtomicInteger();
+
+        failWith429(retryAfterSeconds, requests);
+
+        // Verify that only one attempt was made
+        assertThat(requests.get(), is(1));
+    }
+
+    @Test
+    public void failOn429_RetryLimitExceeded()
+            throws Exception
+    {
+        client = TDClient.newBuilder()
+                .setRetryMaxIntervalMillis(Integer.MAX_VALUE)
+                .setRetryLimit(3)
+                .build()
+                .httpClient;
+
+        final long retryAfterSeconds = 1;
+        final AtomicInteger requests = new AtomicInteger();
+
+        failWith429(retryAfterSeconds, requests);
+
+        // Verify that 4 attempts were made (original request + three retries)
+        assertThat(requests.get(), is(4));
+    }
+
+    private void failWith429(final long retryAfterSeconds, final AtomicInteger requests)
+    {
+        final TDApiRequest req = TDApiRequest.Builder.GET("/v3/system/server_status").build();
+
+        try {
+            ContentResponse resp = client.submitRequest(req, Optional.<String>absent(), new TDHttpClient.DefaultContentHandler()
+            {
+                @Override
+                public ContentResponse submit(Request request)
+                        throws InterruptedException, ExecutionException, TimeoutException
+                {
+                    requests.incrementAndGet();
+                    List<Response.ResponseListener> listeners = ImmutableList.of();
+                    HttpResponse response = new HttpResponse(request, listeners)
+                            .status(429);
+                    response.getHeaders().add("Retry-After", Long.toString(retryAfterSeconds));
```

Comment:
How about adding a test that doesn't add `Retry-After` header to status code 429?